pkg/cri/resource-manager/policy/builtin/balloons/cputree.go

```diff
@@ -0,0 +1,468 @@
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+)
+
+type CPUTopologyLevel int
+
+const (
+	CPUTopologyLevelUndefined CPUTopologyLevel = iota
+	CPUTopologyLevelSystem
+	CPUTopologyLevelPackage
+	CPUTopologyLevelDie
+	CPUTopologyLevelNuma
+	CPUTopologyLevelCore
+	CPUTopologyLevelThread
+	CPUTopologyLevelCount
+)
+
+// cpuTreeNode is a node in the CPU tree.
+type cpuTreeNode struct {
+	name     string
+	level    CPUTopologyLevel
+	parent   *cpuTreeNode
+	children []*cpuTreeNode
+	cpus     cpuset.CPUSet // union of CPUs of child nodes
+
+}
+
+// cpuTreeNodeAttributes contains various attributes of a CPU tree
+// node. When allocating or releasing CPUs, all CPU tree nodes in
+// which allocating/releasing could be possible are stored to the same
+// slice with these attributes. The attributes contain all necessary
+// information for comparing which nodes are the best choices for
+// allocating/releasing, thus traversing the tree is not needed in the
+// comparison phase.
+type cpuTreeNodeAttributes struct {
+	t                *cpuTreeNode
+	depth            int
+	currentCpus      cpuset.CPUSet
+	freeCpus         cpuset.CPUSet
+	currentCpuCount  int
+	currentCpuCounts []int
+	freeCpuCount     int
+	freeCpuCounts    []int
+}
+
+// cpuTreeAllocator allocates CPUs from the branch of a CPU tree
+// where the "root" node is the topmost CPU of the branch.
+type cpuTreeAllocator struct {
+	options cpuTreeAllocatorOptions
+	root    *cpuTreeNode
+}
+
+// cpuTreeAllocatorOptions contains parameters for the CPU allocator
+// that that selects CPUs from a CPU tree.
+type cpuTreeAllocatorOptions struct {
+	// topologyBalancing true prefers allocating from branches
+	// with most free CPUs (spread allocations), while false is
+	// the opposite (packed allocations).
+	topologyBalancing bool
+}
+
+// Strings returns topology level as a string
+func (ctl CPUTopologyLevel) String() string {
+	s, ok := cpuTopologyLevelToString[ctl]
+	if ok {
+		return s
+	}
+	return fmt.Sprintf("CPUTopologyLevelUnknown(%d)", ctl)
+}
+
+// cpuTopologyLevelToString defines names for all CPU topology levels.
+var cpuTopologyLevelToString = map[CPUTopologyLevel]string{
+	CPUTopologyLevelUndefined: "undefined",
+	CPUTopologyLevelSystem:    "system",
+	CPUTopologyLevelPackage:   "package",
+	CPUTopologyLevelDie:       "die",
+	CPUTopologyLevelNuma:      "numa",
+	CPUTopologyLevelCore:      "core",
+	CPUTopologyLevelThread:    "thread",
+}
+
+// UnmarshalJSON unmarshals a JSON string to CPUTopologyLevel
+func (ctl *CPUTopologyLevel) UnmarshalJSON(b []byte) error {
+	if len(b) == 0 {
+		return fmt.Errorf("invalid CPU topology level")
+	}
+	if len(b) > 1 && b[0] == '"' && b[len(b)-1] == '"' {
+		return ctl.UnmarshalJSON(b[1 : len(b)-1])
+	}
+	if i, err := strconv.Atoi(string(b)); err == nil && i >= int(CPUTopologyLevelUndefined) && i < int(CPUTopologyLevelCount) {
+		*ctl = CPUTopologyLevel(i)
+		return nil
+	}
+	name := strings.ToLower(string(b))
```

Comment:
Why not to simply always unmarshal as a string ? Somehing like:

```golang

// cpuTopologyStringToLevel defines reverse (name to level) mapping for all CPU topology levels.
var cpuTopologyStringToLevel map[string]CPUTopologyLevel]

func initStringToLevelMap() {
    cpuTopologyStringToLevel = make(map[string]CPUTopologyLevel)
    for l, n := range cpuTopologyLevelToString {
        cpuTopologyStringToLevel[n] = l
    }
}

func (ctl CPUTopologyLevel) MarshalJSON() ([]byte, error) {
    return json.Marshal(ctl.String())
}

func (ctl *CPUTopologyLevel) UnmarshalJSON(data []byte) error {
    str := ""
    if err := json.Unmarshal(data, &str); err != nil {
        return fmt.Errorf("failed to unmarshal CPUTopologyLevel: %w", err)
    }
    l, ok := cpuTopologyStringToLevel[str]
    if !ok {
        return fmt.Errorf("unknown CPUTopologyLevel %q", str)
    }
    *ctl = l
    return nil
}

...

func init() {
    initStringToLevelMap()
}
```

Reply:
I think it refuses to unmarshal yaml/json int into a string. I.e. this would not work:
```
level: 2
```

but this would:
```
level: "2"
```

But yes, it's questionable do we want to support `int`s without quotes and do we want to support integer values at all(?)

Reply:
Hmm... but if you provide both the marshaller and unmarshaller then such a situation should not arise. Always marshal to/from string (the logical name), and during unmarshalling as the last step look up the value for the unmarshalled string and use that value.

Reply:
Thanks, marshal/unmarshal fixed to work only with strings.